src/backend/metal/shaders/commands.metal

```diff
@@ -20,38 +20,54 @@ vertex BlitVertexData vs_blit(BlitAttributes in [[stage_in]]) {
     return BlitVertexData { pos, in.src_coords, uint(in.dst_coords.z) };
 }
 
-fragment float4 ps_blit_1d(
+fragment float4 ps_blit_1d_float(
```

Comment:
Looks like we are going to have a lot of shader combinations when we expand to 8-bits, 16-bits etc.. should we eventually attempt to generate these during build or something?

Reply:
The more shaders the less incentive we have to load all of them at build time. Currently, all blit shaders are compiled at build time but the pipelines are lazily created. I think we are safe to continue doing so.

> when we expand to 8-bits, 16-bits etc.. 

This is not necessary. Bit sizes don't matter, only the texture dimensions and blit channels do matter. Plus the depth/stencil thing that is still TODO.

Reply:
Yeah, I mean to generate some of the repetitive actual MSL at build time and compile that, then lazily creating pipelines as we do already. Though it's probably overkill for now because we don't that many variations yet.

>This is not necessary. Bit sizes don't matter, only the texture dimensions and blit channels do matter. Plus the depth/stencil thing that is still TODO.

Don't we still have the issue of out-of-bounds writes if we only have `uint`, `int` and `float`? Because even with masking we still can't fill the last element correctly if it's not aligned to 4 bytes (in the existing shaders). For example, if we want to do a copy of 8-bit texture formats like R8, don't we need other shaders to sample `uchar`?

Reply:
I agree there is a bit of redundancy, and it could technically be auto-generated. Let's see if it's going to be actually needed.

> For example, if we want to do a copy of 8-bit texture formats like R8, don't we need other shaders to sample uchar?

Well, for one, I have Vulkan CTS backing up this code, so we are doing fine there :)
My understanding is that the logic of converting from uchar to uint in this case lies on the hands of HW texture sampler. There is no need to generate a different pipeline/shader between u8 and u16 textures.

As for the render targets, the micro-code for converting from shader-output float4/uint4/int4 to the actual format is compiled at the pipeline creation, which knows about the target formats, and that's for we have a pipeline hash map here.

Reply:
> Well, for one, I have Vulkan CTS backing up this code, so we are doing fine there :)
My understanding is that the logic of converting from uchar to uint in this case lies on the hands of HW texture sampler. There is no need to generate a different pipeline/shader between u8 and u16 textures.

Ok, interesting. It just seems strange to me that we need to distinguish `uint` and `int` but not between various sizes.